### ISSUE TITLE
Apply dependencies constraints to deploy configuration of ear plugin

### ISSUE THREAD

**Issue by claudiorossetto** [gradle/gradle#18129]
### Expected Behavior

Suppose i have declared a platform mycompany-platform like this:

```
plugins {
    id 'java-platform'
    id 'maven-publish'
}

group = "com.mycompany"
version = "1.0.0"

dependencies {
    constraints {
        api 'com.mycompany:myproject-ejb:1.+'
        api 'log4j:log4j:1.2.17!!'
    }
}
```

Using the platform in a ear project, should apply the constraints and resolve the correct versions for all libraries:

```
plugins {
    id 'ear'
}

dependencies {

    earlib platform('com.mycompany:mycompany-platform:latest.release')
    earlib 'log4j:log4j'

    earlib platform('com.mycompany:mycompany-platform:latest.release')
    deploy 'com.mycompany:myproject-ejb'

}
```

### Current Behavior

This works for the earlib configuration, but not for the deploy configuration.
The reported error is:

```
Could not determine the dependencies of task ':myproject-ear:ear'.
> Could not resolve all task dependencies for configuration ':myproject-ear:deploy'.
   > Could not find com.mycompany:myproject-ejb:.
     Required by:
         project :myproject-ear
```

As already noted here #11030, this happens because deploy is not transitive.
Should be useful if the constraints were applied to non-transitive configurations, just for the direct dependencies.

### Context

I'm trying to build an ear without having to specify versions.
I'm using gradle version 6.8.2

**Comment by dgeissl** [comment]
This is still relevant.

**Comment by Dalamaar** [comment]
We would like to see this request fixed as well.

Unfortunately I have not yet found a way to access the versions declared in the platform-subproject. Therefore we declare the versions used in the deploy-configuration in the gradle.properties. In the platform-project and for the deploy-configuration we get the version by project.property("version"). It works, but is most certainly not nice.

**Comment by h0tk3y** [comment]
Sorry for the late reply.

This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

***

This issue has a workaround and needs decision regarding introducing the change in the EAR plugin.

**Comment by dgeissl** [comment]
Technically this does not seem to be limited to the Ear plugin, instead it (at least used to be) a general behavior of all non transitive configurations as it already has been mentioned in https://github.com/gradle/gradle/issues/11030.

However you might argue that deploy dependencies should anyways be transitive (which I think they should be) then I'd agree, it can be solved just for the ear plugin.

**Comment by big-guy** [comment]
We don't plan on addressing issues with the EAR Plugin soon, but there may be an alternative way of implementing the deploy feature to not rely on a non-transitive configuration.

**Comment by dgeissl** [comment]
But the alternative way to implement the deploy configuration you teased still needs to be addressed in the EAR Plugin, right?
And this will not be done soon considering the issue is 4 years old?

**Comment by big-guy** [comment]
Correct, this is on our backlog, but we're not focusing on the ear plugin.
